eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/EnforceFinalFieldsCheck.java

```diff
@@ -0,0 +1,248 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.tools.checkstyle.checks;
+
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Verify the whenever a field is assigned just once in constructor to be final
+ * Tree traversal will pre-compute and fill 3 private containers:
+ * nonFinalFields: keep an array of non private fields as tokens (to keep line number)
+ * assignmentsFromConstructor: Save a set of string for each field name that gets its value assigned in constructor
+ * assignmentsFromMethods: Save a set of strings for each field name that gets updated in any method
+ *
+ * On finish tree, check what non-final fields get a value only in constructor and nowhere else by looking for
+ * strings inside nonFinalFields AND assignmentsFromConstructor but NOT in assignmentsFromMethods
+ */
+public class EnforceFinalFieldsCheck extends AbstractCheck {
+    private static final String ERROR_SUGGESTION = "You should consider making the field final, " +
```

Comment:
Run -Dinclude-non-shipping-modules. This is probably going to result in a build failure.